File: RecursionBacktracking/factorial.py
# Here we assume that number is always +ve or 0(Non negative)
def factorial(num):
    if num == 0:
        return 1
    return num * factorial(num-1)

# time complexity in worst case O(N) and
# since it is recursion space complexity also O(N) becoz of call stack_Queue

# memoization(usually fast check using time import)
def fact(num):
    fact=[1, 1]
    for i in range(2, num+1):
        fact.append(fact[i-1] * i)
    return fact[num]

if __name__ == "__main__":
     print(f"{factorial(int(input()))}")
     print(fact(int(input())))

#task here to minimize it < O(N)

File: RecursionBacktracking/factorial.cpp
/*
time complexity : O(number) -> there are n calls.
space complexity: O(number) due to recursion stack
*/

#include <iostream>
using namespace std;
int factorial(int number){

    //base case 
    if (number == 0) return 1;

    //recursive case
    int fact = number * factorial(number-1);
    return fact;
}

int main(){

    int number;
    cin >> number;

    cout << "factorial of " << number << " is: " << factorial(number) << endl;
    return 0;
}